### ISSUE TITLE
out of range violation access with hist tree method

### ISSUE THREAD

**Issue by DoanHuuJacques** [dmlc/xgboost#4042]
When running in debug mode XGBoost on Windows with hist as tree method, an access violation has been detected at this location (column_matrix.h)
See the comment with access violation indication

`  inline Column<T> GetColumn(unsigned fid) const {
    const bool valid_type = std::is_same<T, uint32_t>::value
                          || std::is_same<T, uint16_t>::value
                          || std::is_same<T, uint8_t>::value;
    CHECK(valid_type);
    Column<T> c;

    c.type = type_[fid];
    const size_t block_offset = boundary_[fid].index_begin / packing_factor_;
    const size_t elem_offset = boundary_[fid].index_begin % packing_factor_;
    c.index = reinterpret_cast<const T*>(&index_[block_offset]) + elem_offset;
    c.index_base = index_base_[fid];
    c.row_ind = &row_ind_[boundary_[fid].row_ind_begin];  <== row_ind vector is empty (access violation)
    c.len = boundary_[fid].index_end - boundary_[fid].index_begin;

    return c;
  }`

Indeed, row_ind_ vector is empty in case of dense column type
I would say to modify as follows
if ( c.type == ColumnType::kSparseColumn)
  c.row_ind = &row_ind_[boundary_[fid].row_ind_begin];
else
 c.row_ind = nullptr;

Basically, it's impossible to debug on Windows with debug XGBoost binary (this access violation is blocking)

**Comment by RAMitchell** [comment]
I thought I fixed this. Is this the latest version from source?

**Comment by DoanHuuJacques** [comment]
@RAMitchell 
Hi
this header file has been refactored but the access violation is still there

    Column c(type_[fid], &index_[boundary_[fid].index_begin], index_base_[fid],
             &row_ind_[boundary_[fid].row_ind_begin],
             boundary_[fid].index_end - boundary_[fid].index_begin);
    return c;
  }

It's better to protect data access with below condition when row_ind_ is empty:
type_[fid] == ColumnType::kSparseColumn) ? &row_ind_[boundary_[fid].row_ind_begin] : nullptr

**Comment by DoanHuuJacques** [comment]
@RAMitchell 
I will do
Thanks
